### ISSUE TITLE
Schema specific fields

### ISSUE THREAD

**Issue by stevelacey** [rebing/graphql-laravel#603]
Schema specific fields solve a problem I had when I wanted to make changes to existing types for new schema without creating backwards incompatible changes in existing schema that use them.

Instead of duping the type and adding/removing/changing fields, I wanted to be able to say "this field is deprecated on this schema" and "this field is only available on this one"

I did this by adding an optional DSL on top of the field key names that works similar to Laravel validation rules, using the array key so that they're distinct and don't clobber:

```php
'roles(schema:app,user)' => [
    'type' => Type::listOf(GraphQL::type('Role')),
    'description' => 'The roles assigned to the user.',
],

// Deprecated Fields

'roles(schema:default)' => [
    'type' => Type::listOf(Type::string()),
    'description' => 'The roles assigned to the user.',
    'deprecationReason' => 'Roles are no longer exposed via the default graphql schema',
],
```

The way I went about doing it is by intercepting the `getFields` call and evaluating anything in brackets as a call to a local filter method that adjusts the `$fields` array accordingly:

```php
abstract class GraphQLType extends Type
{
    public function getFields() : array
    {
        $fields = parent::getFields();

        if ($this->enumObject || $this->inputObject) {
            return $fields;
        }

        $fields = $this->filterFields($fields);

        return $fields;
    }

    /**
     * Apply filters to any fields that pass them
     *
     * This is primarily to give a terse DSL for conditionally adding, removing
     * or changing the resolver for fields from certain graphql schema
     *
     * e.g. app_id(schema:default), roles(schema:platform)
     */
    private function filterFields(array $fields)
    {
        $allFields = [];

        foreach ($fields as $name => $field) {
            if (preg_match('/^(\w+)\((.*)\)$/', $name, $matches)) {
                list($_, $name, $filters) = $matches;

                foreach (explode('|', $filters) as $filter) {
                    $filter = explode(':', $filter);

                    $filterName = "{$filter[0]}Filter";
                    $filterArgs = isset($filter[1]) ? explode(',', $filter[1]) : [];

                    if ($this->{$filterName}($filterArgs)) {
                        $allFields[$name] = $field;
                    }
                }

                continue;
            }

            $allFields[$name] = $field;
        }

        return $allFields;
    }

    private function schemaFilter(array $args = []) : bool
    {
        $schema = config('graphql.default_schema');

        if (Route::current()) {
            $schema = Arr::first(Route::current()->parameters()) ?? $schema;
        }

        return in_array($schema, $args);
    }
}
```

I doubt we'd want to merge the key DSL behaviour into core, but perhaps schema specific fields are useful enough to warrant support in some form, this solves a lot of API versioning problems GraphQL creates in large schema in a reasonably elegant and maintainable way.

**Comment by stevelacey** [comment]
I'll share another couple of things you can do by overriding `getFields` just in case it helps anyone else that comes across this issue:

**Field traits:**

Allows you to share field definitions across a bunch of types.

```php
$uses = array_flip(class_uses_recursive(static::class));

if (isset($uses[HasTimestamps::class])) {
    $fields = array_merge($fields, $this->getTimestampFields());
}

if (isset($uses[HasPagination::class])) {
    $fields = array_merge($fields, $this->getPaginationFields());
}
```

**Mass renames:**

Allows you to make schema-wide changes without doing a full rewrite, e.g. by deprecating all snake_case fields and exposing them instead as camelCase.

```php
$fields = $this->deprecateSnakeCaseFields($fields);

// ...

private function deprecateSnakeCaseFields(array $fields)
{
    $allFields = [];

    foreach ($fields as $name => $field) {
        $allFields[$name] = $field;

        if (
            Str::contains($name, '_')
            && !isset($fields[$camelcase = Str::camel($name)])
            && !isset($fields[$snakecase = $name]['deprecationReason'])
        ) {
            $allFields[$camelcase] = $field;
            $allFields[$camelcase]['name'] = $camelcase;
            $allFields[$camelcase]['resolve'] = $allFields[$camelcase]['resolve'] ?? $this->buildFallbackFieldResolver($snakecase);
            $allFields[$snakecase]['deprecationReason'] = "Renamed to {$camelcase}";
        }
    }

    return $allFields;
}

/**
 * Field resolver for camelCase fields that base off of old snake_case defs
 *
 * Avoids hitting the defaultFieldResolver which will resolve a camelCase
 * model property if there is one available, which might not have been what
 * the old snake_case field was serving.
 */
private function buildFallbackFieldResolver(string $snakecase)
{
    return function ($source, $args, $context, ResolveInfo $info) use ($snakecase) {
        $property = null;

        if (is_array($source) || $source instanceof ArrayAccess) {
            if (isset($source[$snakecase])) {
                $property = $source[$snakecase];
            }
        } elseif (is_object($source)) {
            if (isset($source->{$snakecase})) {
                $property = $source->{$snakecase};
            }
        }

        if ($property instanceof Closure) {
            return $property($source, $args, $context, $info);
        }

        if (is_array($property) && in_array($info->returnType, ['Object', '[Object]'])) {
            return Arr::camel($property);
        }

        return $property;
    };
}
```

**Comment by mfn** [comment]
> we're definitely deep in poweruser territory with these works

I'm not the original author(s) of this library so I can't comment on the initial vision. But I've no problems introducing such features as long as they overall fit into the picture.

I looking at the field name rewrite with some smirk in my face. Reminds of some of many discussions I had with teams regarding automatic renaming behaviour because the convention, whilst being 1:1 mapping what the "backend" produces", didn't fit into the frontends "naming convention". I fought hard not to implement such rewriting layer, I prevailed so far.

I like the schema specific fields but I'm not fond of the syntax. The name should be just the name and the configuration array is in fact there for exactly such reasons.
Yes, there's clobber/pollution issue because right now they're flat shared with the underlying library 😬
But especially with your "calling `getFields` and doing further stuff with it"-example would benefit if then field name would not contain such controlling meta data and rather by just further config options.

On that note for example, I wish all the `SelectFields` (which I personally don't use in any projects) would use its own key and a sub-array for the config.

**Comment by mfn** [comment]
Maybe something could be cooked up by just adding a new `'schema'` attribute, being an array of schemas a specific field is supposed to appear in.

I'm actually less convinced this is a good idea and since https://github.com/rebing/graphql-laravel/pull/658 was recently merged, making a dedicated type should probably be the way to go.
For this to reduce the redundancy, a way could be to:
- create an `AbstracMyType extends \Rebing\GraphQL\Support\Type` and implement `fields()`
- create your `TypeInSchemaOne extends AbstracMyType`, implement `fields()`, call parent, add/remove/etc. fields
- create your `TypeInSchemaTwo extends AbstracMyType`, implement `fields()`, call parent, add/remove/etc. fields
  …

Basically it's already possible now and the argument could be made it could be done more accessible / less verbose.

I'm closing this as it was created as a question and we got already some answers what could be done, I suggest the next move would be a PR to discuss the technical side.

Thanks!
